File: sample/src/main/java/com/afollestad/materialcabsample/MainActivity.java
package com.afollestad.materialcabsample;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.afollestad.materialcab.MaterialCab;

import java.lang.reflect.Field;

/** @author Aidan Follestad (afollestad) */
public class MainActivity extends AppCompatActivity
    implements MainAdapter.Callback, MaterialCab.Callback {

  private MainAdapter adapter;
  private MaterialCab cab;
  private Toast toast;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    adapter = new MainAdapter(this);
    RecyclerView list = (RecyclerView) findViewById(R.id.list);
    list.setLayoutManager(new LinearLayoutManager(this));
    list.setAdapter(adapter);

    setSupportActionBar((Toolbar) findViewById(R.id.main_toolbar));

    if (savedInstanceState != null) {
      cab = MaterialCab.restoreState(savedInstanceState, this, this);
      adapter.restoreState(savedInstanceState);
    } else {
      for (int i = 0; i <= 80; i++) {
        adapter.add("Item " + i);
      }
    }
  }

  @Override
  protected void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    if (cab != null) {
      cab.saveState(outState);
    }
    if (adapter != null) {
      adapter.saveState(outState);
    }
  }

  private void showToast(String text) {
    if (toast != null) {
      toast.cancel();
    }
    toast = Toast.makeText(this, text, Toast.LENGTH_SHORT);
    toast.show();
  }

  @Override
  public void onItemClicked(int index, boolean longClick) {
    if (longClick || (cab != null && cab.isActive())) {
      onIconClicked(index);
      return;
    }
    showToast(adapter.getItem(index));
  }

  @Override
  public void onIconClicked(int index) {
    adapter.toggleSelected(index);
    if (adapter.getSelectedCount() == 0) {
      cab.finish();
      return;
    }
    if (cab == null) {
      cab = new MaterialCab(this, R.id.cab_stub).start(this);
    } else if (!cab.isActive()) {
      cab.reset().start(this);
    }
    cab.title(getString(R.string.x_selected, adapter.getSelectedCount()));
  }

  @Override
  public boolean onCabCreated(@NonNull MaterialCab cab, Menu menu) {
    // Makes the icons in the overflow menu visible
    if (menu.getClass().getSimpleName().equals("MenuBuilder")) {
      try {
        Field field = menu.getClass().getDeclaredField("mOptionalIconsVisible");
        field.setAccessible(true);
        field.setBoolean(menu, true);
      } catch (Exception ignored) {
        ignored.printStackTrace();
      }
    }
    return true; // allow creation
  }

  @Override
  public boolean onCabItemClicked(@NonNull MenuItem item) {
    showToast((String) item.getTitle());
    return true;
  }

  @Override
  public boolean onCabFinished(@NonNull MaterialCab cab) {
    adapter.clearSelected();
    return true; // allow destruction
  }

  @Override
  public void onBackPressed() {
    if (cab != null && cab.isActive()) {
      cab.finish();
      cab = null;
    } else {
      super.onBackPressed();
    }
  }
}

File: sample/src/main/java/com/afollestad/materialcabsample/MainActivity.kt
/**
 * Designed and developed by Aidan Follestad (@afollestad)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.afollestad.materialcabsample

import android.os.Bundle
import android.view.Menu
import androidx.appcompat.app.AppCompatActivity
import com.afollestad.materialcab.attached.AttachedCab
import com.afollestad.materialcab.attached.destroy
import com.afollestad.materialcab.attached.isActive
import com.afollestad.materialcab.createCab
import com.afollestad.recyclical.datasource.emptySelectableDataSource
import com.afollestad.recyclical.setup
import com.afollestad.recyclical.viewholder.isSelected
import com.afollestad.recyclical.withItem
import kotlinx.android.synthetic.main.activity_main.list

/** @author Aidan Follestad (afollestad) */
class MainActivity : AppCompatActivity() {
  private val dataSource = emptySelectableDataSource().apply {
    onSelectionChange { invalidateCab() }
  }
  private var mainCab: AttachedCab? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    setSupportActionBar(findViewById(R.id.main_toolbar))

    dataSource.set(
        IntArray(100) { it + 1 }
            .map { MainItem("Item #$it") }
    )

    list.setup {
      withDataSource(dataSource)
      withItem<MainItem, MainViewHolder>(R.layout.listitem_main) {
        onBind(::MainViewHolder) { index, item ->
          itemView.isActivated = isSelected()
          title.text = item.title
          icon.setOnClickListener {
            dataSource.toggleSelectionAt(index)
          }
        }
        onClick {
          if (hasSelection()) {
            toggleSelection()
          } else {
            toast("Clicked $item")
          }
        }
        onLongClick { toggleSelection() }
      }
    }
  }

  private fun invalidateCab() {
    if (!dataSource.hasSelection()) {
      mainCab?.destroy()
      return
    }

    if (mainCab.isActive()) {
      mainCab?.apply {
        title(literal = getString(R.string.x_selected, dataSource.getSelectionCount()))
      }
    } else {
      mainCab = createCab(R.id.cab_stub) {
        title(literal = getString(R.string.x_selected, dataSource.getSelectionCount()))
        menu(R.menu.menu_cab)
        popupTheme(R.style.ThemeOverlay_AppCompat_Light)
        slideDown()

        onCreate { _, menu -> onCabCreated(menu) }
        onSelection {
          toast(it.title as String)
          true
        }
        onDestroy {
          dataSource.deselectAll()
          true
        }
      }
    }
  }

  private fun onCabCreated(menu: Menu): Boolean {
    // Makes the icons in the overflow menu visible
    if (menu.javaClass.simpleName == "MenuBuilder") {
      try {
        val field = menu.javaClass.getDeclaredField("mOptionalIconsVisible")
        field.isAccessible = true
        field.setBoolean(menu, true)
      } catch (ignored: Exception) {
        ignored.printStackTrace()
      }
    }
    return true // allow creation
  }

  override fun onBackPressed() {
    if (!mainCab.destroy()) {
      super.onBackPressed()
    }
  }
}
